luarules/gadgets/cmd_engagement_range.lua

```diff
@@ -0,0 +1,348 @@
+
+local unitDefaultEngageRangeOverride = {
+	legbunk = WeaponDefNames.legbunk_railgunt2 and WeaponDefNames.legbunk_railgunt2.range or nil,
+}
+
+--------------------------------------------------------------------------------
+-- Global variables ------------------------------------------------------------
+
+local ARMORTYPE_DEFAULT = Game.armorTypes.default
+local ARMORTYPE_VTOL = Game.armorTypes.vtol
+
+local CMD_MOVE_STATE = CMD.MOVE_STATE
+local CMD_ENGAGE_STATE = GameCMD.ENGAGE_STATE
+
+--------------------------------------------------------------------------------
+-- Local variables -------------------------------------------------------------
+
+---Units can have their own params but share this base table in their cmdDescs.
+---@type CommandDescription
+local engageStateCmdDesc = {
+	id       = CMD_ENGAGE_STATE,
+	type     = CMDTYPE.ICON_MODE,
+	name     = "engage_state",
+	action   = "engage_state",
+	tooltip  = "Cycle between weapon engage ranges",
+	cursor   = "cursornormal",
+	queueing = false,
+}
+
+local ENGAGESTATE_DEFAULT = 0
+
+-- Labels support mixing and matching to create param sets but add complexity.
+local labels = {
+	default   = "Engage Default",
+	close     = "Engage Close Range",
+	secondary = "Engage Secondary",
+}
+
+for _, label in ipairs({ "default", "close", "secondary" }) do
+	labels[labels[label]] = label -- reverse lookup
+end
+
+-- This is specifically to offset the CMD_FIGHT leash.
+-- Different commands may use a different leash radius.
+local moveStateLeashRadius = {
+	[CMD.MOVESTATE_NONE]     = 0,
+	[CMD.MOVESTATE_HOLDPOS]  = 0,
+	[CMD.MOVESTATE_MANEUVER] = 100,
+	[CMD.MOVESTATE_ROAM]     = 400,
+}
+
+-- Create up to three pseudo-weapons per unit:
+-- - The default engagement range weapon.
+-- - The close-range engagement weapon.
+-- - The secondary-role engagement weapon.
+local unitEngageRangeInfo = {}
+
+local function ignoreWeapon(weaponDef)
+	local custom = weaponDef.customParams
+
+	return custom.bogus or
+		-- These cases are redundant with the `bogus` manual category:
+		custom.nofire or -- kicks, stomps, prime/arm/launch effects
+		weaponDef.range < 10 or -- self-detonators, fake aim points
+		(
+			-- non-damaging with no valid secondary effect
+			weaponDef.damages[ARMORTYPE_DEFAULT] <= 1 and
+			weaponDef.damages[ARMORTYPE_VTOL] <= 1 and
+			not weaponDef.name:find("juno_pulse") and -- hardcoded :/ rip
+			not (custom.carried_unit and tonumber(custom.maxunits or 1) >= 1)
+		) or
+		-- This is a true weapon but is not used for automatic ranging:
+		weaponDef.commandfire
+end
+
+local function getArmorDamages(weaponDef)
+	local damages = weaponDef.damages
+	local damageDefault = damages[ARMORTYPE_DEFAULT] or 0
+	local damageAntiAir = damages[ARMORTYPE_VTOL] or 0
+	local damageMax = math.max(damageDefault, damageAntiAir, 0)
+
+	local damageRate = damageMax * (weaponDef.burst or 1) * (weaponDef.projectiles or 1) /
```

Comment:
I don't believe either of the two are supposed to ever be nil
```suggestion
	local damageRate = damageMax * weaponDef.burst * weaponDef.projectiles /
```